src/shared/components/query/CancerStudySelector.tsx

```diff
@@ -0,0 +1,267 @@
+import {memoizeWith} from "../../api/memoize";
+
+export const CANCER_TYPE_ROOT = 'tissue';
+type Node = TypeOfCancer|CancerStudy;
+function isCancerType(node:TypeOfCancer|CancerStudy) {
+	return !(node as CancerStudy).cancerStudyIdentifier;
+}
+
+export interface ICancerStudySelectorProps
+{
+	style?: React.CSSProperties;
+
+	cancerTypes: TypeOfCancer[];
+	studies: CancerStudy[];
+
+	selectedCancerType?: string;
+	selectedStudies?: string[];
+
+	onStateChange?: (newState:ICancerStudySelectorState) => void;
+}
+
+export interface ICancerStudySelectorState
+{
+	selectedCancerType?: string;
+	selectedStudies?: string[];
+}
+
+export default class CancerStudySelector extends React.Component<ICancerStudySelectorProps, ICancerStudySelectorState>
+{
+	constructor(props: ICancerStudySelectorProps)
+	{
+		super(props);
+		this.state = {};
+	}
+
+	componentDidMount()
+	{
+	}
+
+	@memoizeWith(function() { return [this.props.cancerTypes, this.props.studies] })
+	getTreeInfo() {
+		let {cancerTypes, studies} = this.props;
+    	let dict_cancerTypeId_cancerType:Dictionary<TypeOfCancer|undefined> = _.keyBy(cancerTypes, cancerType => cancerType.typeOfCancerId);
+    	let dict_cancerTypeId_childCancerTypes:Dictionary<TypeOfCancer[]|undefined> = _.groupBy(cancerTypes, cancerType => cancerType.parent);
+    	let dict_cancerTypeId_childStudies:Dictionary<CancerStudy[]|undefined> = _.groupBy(studies, study => study.typeOfCancerId);
+    	let	dict_cancerTypeId_flatStudiesList:Dictionary<CancerStudy[]> = {};
+    	let dict_studyId_study:Dictionary<CancerStudy|undefined> = _.keyBy(studies, study => study.cancerStudyIdentifier);
+
+		let rootCancerType:TypeOfCancer = {
+			clinicalTrialKeywords: '',
+			dedicatedColor: '',
+			name: 'All',
+			parent: '',
+			shortName: 'All',
+			typeOfCancerId: CANCER_TYPE_ROOT
+		};
+		dict_cancerTypeId_cancerType[CANCER_TYPE_ROOT] = rootCancerType;
+
+    	function getFlatStudiesList(cancerTypeId:string, output:CancerStudy[]) {
+    		let childCancerTypes = dict_cancerTypeId_childCancerTypes[cancerTypeId];
+    		if (childCancerTypes)
+    			for (let child of childCancerTypes)
+    				getFlatStudiesList(child.typeOfCancerId, output);
+    		let childStudies = dict_cancerTypeId_childStudies[cancerTypeId];
+    		if (childStudies)
+				output.push.apply(output, childStudies);
+			return output;
+		}
+
+		for (let cancerTypeId in dict_cancerTypeId_cancerType)
+			dict_cancerTypeId_flatStudiesList[cancerTypeId] = getFlatStudiesList(cancerTypeId, []);
+
+		return {
+			rootCancerType,
+			dict_cancerTypeId_cancerType,
+			dict_cancerTypeId_childCancerTypes,
+			dict_cancerTypeId_childStudies,
+			dict_cancerTypeId_flatStudiesList,
+			dict_studyId_study,
+		};
+	}
+
+	@memoizeWith(function() { return [this.props.cancerTypes, this.props.studies] })
+    getTreeDescriptor():ITreeDescriptor<TypeOfCancer|CancerStudy> & {rootNode: TypeOfCancer, setExpanded: (node:Node, value:boolean) => void} {
+		let {

```

Comment:
some inconsistent indentation in this file.  that is one linting task i think is important.  otherwise things get very unreadable fast.  it's important that people be able to use the IDE reformat and that it aligns with linting